cli/flox/src/commands/services/mod.rs

```diff
@@ -206,6 +304,36 @@ pub(crate) fn service_does_not_exist_error(name: &str) -> anyhow::Error {
     anyhow!(format!("Service '{name}' not found."))
 }
 
+/// Add more information to connection errors when talking to the service manager.
+///
+/// Process-Compose commands may fail with [LoggedError::SocketDoesntExist]
+/// if the specific socket doesn't exist,
+/// or if the service manager has quit without cleaning up the socket,
+/// i.e. the socket is unresponsive.
+///
+/// This function adds more context to the error in the latter case.
+///
+/// For practical purposes, we apply this to the [ProcessStates::read] call in
+/// all services commands, as that is typically the first call that can fail.
+/// For following commands, we can assume that the service manager is running.
+///
+/// TODO: we might move this into a `processes` method on [ServicesEnvironment],
+/// as that would bring error handling closer to the environment (i.e the origin of the `socket`).
+pub(super) fn handle_service_connection_error(error: ServiceError, socket: &Path) -> anyhow::Error {
```

Comment:
> nonblocking: I'm not completely sure if there's actually a simpler way, but it still feels a bit backwards to me that we call ProcessStates::read, and then we check if there's a socket afterward to decide what error there is.

+1 for combining this with `ProcessStates::read`. We don't currently call `read` without `handle_service_connection_error`; do we foresee any reason to do so in the near future?

> And it seems like using socket.exists() to mean we think process-compose is up and running can lead to potentially bad states (e.g. we never start a new process-compose because an old one has crashed leaving behind the socket).

Yeah, it could:
```
flox [tmp dcarley/shell (remote)] 
(3) ~/projects/flox/flox on HEAD (5c07f6fa) [$]
% pkill -9 process-compose      
flox [tmp dcarley/shell (remote)] 
(3) ~/projects/flox/flox on HEAD (5c07f6fa) [$]
% ft services restart     
❌ ERROR: unknown error: Get \"http://unix/processes\": dial unix /Users/dcarley/.cache/flox/run/flox.ee4e7845.sock: connect: connection refused
```

Although we have other issues in that situation, like not being able to (re)start the orphaned services it has left behind because they're still holding ports or files open. Some of that may be addressed by https://github.com/flox/flox/issues/1613 but I think we could follow-up with an additional check to see if `process-compose` is functional sometime between this PR and that issue.

Reply:
> +1 for combining this with ProcessStates::read. We don't currently call read without handle_service_connection_error; do we foresee any reason to do so in the near future?

I think i wouldn;t change `ProcessStates::read` for it but have methods alongside either on `ProcessStates` or `ServicesEnvironment` that does this.
I'm still a bit unsure if we should really concentrate this on `ProcessStates::read` at all or whether we should check for socket existence in all calls to process compose (rather than relying on the somewhat ambiguous logged error form process compose).

Reply:
Agree, it should probably be applied to all `process-compose` calls, rather than just affecting the commands that happen to depend on needing `ProcessStates` first. That sounds like it would have the same effect as I was asking for, which is that all of the callers shouldn't need to know that they also need to call `handle_service_connection_error` if we don't have any cases where that's not currently required.

Reply:
I meved some things around in [099aefc](https://github.com/flox/flox/pull/1969/commits/099aefcc4694431b5de8ae7c0a50e86c453a6ae4) including, adding `ServicesEnvironment::processes_if_expected_running`.

I tried to keep the changes to activate minimal for this PR, as i feel a bit like an elephant in the glass house there.

The wholistic approach to this would probably be to add errors to `ServiceError` and proactively check for socket existence for all process_compose calls (except `up` i guess?) and map those errors to meaningful actions in the cli.
However that likely needs some restructuring in the provider, which i tried to keep mostly untouched for now to keep the changes small

Reply:
what i'm thinking might be easier still, and would require mixing up env and services that much is, just adding an explicit guard method that checks if 1) the socket exists and 2) whether the socket is alive by connecting to it (or running status and discarding the result)

And then from there on just use `ProcessStates::read()` and friends